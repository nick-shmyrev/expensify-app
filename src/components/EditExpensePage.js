import React from 'react';
import { connect } from 'react-redux';
import { editExpense, removeExpense } from '../actions/expenses-actions';
import ExpenseForm from './ExpenseForm';

export class EditExpensePage extends React.Component {
  onExpenseFormSubmit = expense => {
    this.props.editExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  };
  
  handleRemoveExpense = (e) => {
    e.preventDefault();
    this.props.removeExpense(this.props.expense.id);
    this.props.history.push('/');
  };
  
  render() {
    return (
      <div>
        <ExpenseForm
          expense={this.props.expense}
          onExpenseFormSubmit={this.onExpenseFormSubmit}
        />
        <button
          name="removeExpense"
          onClick={this.handleRemoveExpense}
        >
          Remove
        </button>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  expense: state.expenses.find(el => el.id === props.match.params.id),
});
const mapDispatchToProps = dispatch => ({
  editExpense: (id, update) => dispatch(editExpense(id, update)),
  removeExpense: id => dispatch(removeExpense(id)),
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
