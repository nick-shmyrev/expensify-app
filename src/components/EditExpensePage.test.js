import React from 'react';
import { shallow } from 'enzyme';
import { EditExpensePage } from './EditExpensePage';

describe('EditExpensePage component', () => {
  let history;
  let wrapper;
  let editExpense;
  let removeExpense;
  const expense = {
    id: '023-abc',
    description: 'Description',
    amount: 39.5,
    note: 'A note',
    createdAt: 0,
  };
  
  beforeEach(() => {
    editExpense = jest.fn();
    removeExpense = jest.fn();
    history = { push: jest.fn() };
    wrapper = shallow(<EditExpensePage expense={expense} editExpense={editExpense} removeExpense={removeExpense} history={history}/>);
  });
  
  test('should render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
  
  test('should handle editExpense', () => {
    wrapper.find('ExpenseForm').prop('onExpenseFormSubmit')(expense);
  
    expect(history.push).toHaveBeenLastCalledWith('/');
    expect(editExpense).toHaveBeenLastCalledWith(expense.id, expense);
  });
  
  test('should handle removeExpense', () => {
    wrapper.find('button[name="removeExpense"]').simulate('click', {
      preventDefault: () => {},
    });
    
    expect(removeExpense).toHaveBeenLastCalledWith(expense.id);
    expect(history.push).toHaveBeenLastCalledWith('/');
  });
});
