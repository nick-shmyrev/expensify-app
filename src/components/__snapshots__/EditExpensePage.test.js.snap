// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditExpensePage component should render correctly 1`] = `
<div>
  <ExpenseForm
    expense={
      Object {
        "amount": 39.5,
        "createdAt": 0,
        "description": "Description",
        "id": "023-abc",
        "note": "A note",
      }
    }
    onExpenseFormSubmit={[Function]}
  />
  <button
    name="removeExpense"
    onClick={[Function]}
  >
    Remove
  </button>
</div>
`;
